File: auto-translate-uk-to-pt.py
#!/usr/bin/env python3
"""
Auto-translate Ukrainian text to Portuguese (pt-PT) in client, tech, dispatcher panels.
Uses Gemini AI API.
"""

import os
import re
import json
import time
import glob
import google.generativeai as genai
from dotenv import load_dotenv

load_dotenv()

GEMINI_API_KEY = os.environ.get("GEMINI_API_KEY")
genai.configure(api_key=GEMINI_API_KEY)
model = genai.GenerativeModel("gemini-2.0-flash")

UK_PATTERN = re.compile(r'[\u0400-\u04FF]+')

TARGET_DIRS = [
    "pages/client",
    "pages/tech",
    "pages/dispatcher",
]

def has_ukrainian(text):
    return bool(UK_PATTERN.search(text))

def extract_ukrainian_segments(text):
    """Extract all segments that contain Ukrainian characters (word boundaries)."""
    # Find all ukrainian-containing words with surrounding context (up to full token)
    segments = set()
    # Extract full "words" or tokens containing Ukrainian
    for match in re.finditer(r'[^\s<>"\'=\{\}]+', text):
        token = match.group()
        if has_ukrainian(token):
            segments.add(token)
    return segments

def collect_files(dirs, extensions=('.html', '.js')):
    files = []
    for d in dirs:
        for ext in extensions:
            files.extend(glob.glob(f"{d}/**/*{ext}", recursive=True))
            files.extend(glob.glob(f"{d}/*{ext}"))
    return list(set(files))

def translate_batch(texts):
    """Translate a list of Ukrainian texts to Portuguese using Gemini."""
    if not texts:
        return {}
    
    items = "\n".join([f"{i+1}. {t}" for i, t in enumerate(texts)])
    prompt = f"""Translate the following Ukrainian words/phrases to European Portuguese (pt-PT).
These are from a web application about elevator maintenance management.
Return ONLY a JSON object mapping each original text to its Portuguese translation.
Keep technical terms, HTML attributes, variable names unchanged.
Do not translate: class names, IDs, variable names, function names, URLs, numbers.
Only translate human-readable Ukrainian text.

Texts to translate:
{items}

Return format: {{"original": "translation", ...}}"""

    try:
        response = model.generate_content(prompt)
        resp_text = response.text.strip()
        # Extract JSON from response
        json_match = re.search(r'\{.*\}', resp_text, re.DOTALL)
        if json_match:
            return json.loads(json_match.group())
    except Exception as e:
        print(f"  ⚠️ Gemini error: {e}")
    return {}

def process_file(filepath, translations_cache):
    """Process a single file - find Ukrainian text and replace with Portuguese."""
    try:
        with open(filepath, 'r', encoding='utf-8') as f:
            content = f.read()
    except Exception as e:
        print(f"  ❌ Cannot read {filepath}: {e}")
        return False, 0

    if not has_ukrainian(content):
        return False, 0

    # Collect all Ukrainian-containing segments
    segments = extract_ukrainian_segments(content)
    
    # Filter out ones we haven't translated yet
    to_translate = [s for s in segments if s not in translations_cache]
    
    if to_translate:
        print(f"  🔄 Translating {len(to_translate)} new segments...")
        # Batch in groups of 30
        batch_size = 30
        for i in range(0, len(to_translate), batch_size):
            batch = to_translate[i:i+batch_size]
            new_translations = translate_batch(batch)
            translations_cache.update(new_translations)
            if len(to_translate) > batch_size:
                time.sleep(1)  # Rate limit

    # Apply translations to content
    new_content = content
    changes = 0
    
    # Sort by length (longest first) to avoid partial replacements
    for original in sorted(segments, key=len, reverse=True):
        if original in translations_cache:
            translation = translations_cache[original]
            if translation and translation != original and not has_ukrainian(translation):
                if original in new_content:
                    new_content = new_content.replace(original, translation)
                    changes += 1

    if changes > 0 and new_content != content:
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(new_content)
        return True, changes
    
    return False, 0

def main():
    print("🇵🇹 Auto-translate Ukrainian → Portuguese (pt-PT)")
    print("=" * 60)
    
    files = collect_files(TARGET_DIRS)
    print(f"📁 Found {len(files)} files to process")
    
    # Load existing cache if any
    cache_file = ".translation-cache.json"
    translations_cache = {}
    if os.path.exists(cache_file):
        with open(cache_file, 'r', encoding='utf-8') as f:
            translations_cache = json.load(f)
        print(f"💾 Loaded {len(translations_cache)} cached translations")
    
    total_changed = 0
    total_changes = 0
    
    for filepath in sorted(files):
        rel_path = filepath.replace('/home/andriy/deapseak/', '')
        
        # Quick check
        try:
            with open(filepath, 'r', encoding='utf-8') as f:
                content = f.read()
            if not has_ukrainian(content):
                continue
        except:
            continue
        
        print(f"\n📄 {rel_path}")
        changed, changes = process_file(filepath, translations_cache)
        
        if changed:
            print(f"  ✅ Applied {changes} replacements")
            total_changed += 1
            total_changes += changes
        else:
            print(f"  ℹ️  No changes applied")
        
        # Save cache periodically
        with open(cache_file, 'w', encoding='utf-8') as f:
            json.dump(translations_cache, f, ensure_ascii=False, indent=2)
    
    print(f"\n{'=' * 60}")
    print(f"✅ Done! Modified {total_changed} files, {total_changes} replacements")
    print(f"💾 Cache saved: {len(translations_cache)} translations")

if __name__ == "__main__":
    main()

File: auto-translate-uk-to-pt.js
/**
 * Auto-translate Ukrainian text to Portuguese (pt-PT)
 * in client, tech, dispatcher panels using Gemini AI.
 */

require('dotenv').config();
const { GoogleGenerativeAI } = require('@google/generative-ai');
const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

const genai = new GoogleGenerativeAI(process.env.GEMINI_API_KEY);
const model = genai.getGenerativeModel({ model: 'gemini-2.0-flash' });

const UK_REGEX = /[\u0400-\u04FF]/;
const TARGET_DIRS = ['pages/client', 'pages/tech', 'pages/dispatcher'];
const CACHE_FILE = '.translation-cache.json';
const EXTENSIONS = ['.html', '.js'];

// Load cache
let cache = {};
if (fs.existsSync(CACHE_FILE)) {
    cache = JSON.parse(fs.readFileSync(CACHE_FILE, 'utf-8'));
    console.log(`💾 Loaded ${Object.keys(cache).length} cached translations`);
}

function hasUkrainian(text) {
    return UK_REGEX.test(text);
}

function saveCache() {
    fs.writeFileSync(CACHE_FILE, JSON.stringify(cache, null, 2), 'utf-8');
}

function collectFiles(dirs) {
    const files = [];
    for (const dir of dirs) {
        if (!fs.existsSync(dir)) continue;
        const walk = (d) => {
            for (const entry of fs.readdirSync(d, { withFileTypes: true })) {
                const full = path.join(d, entry.name);
                if (entry.isDirectory()) walk(full);
                else if (EXTENSIONS.includes(path.extname(entry.name))) {
                    files.push(full);
                }
            }
        };
        walk(dir);
    }
    return files;
}

/**
 * Extract Ukrainian-containing line segments for translation.
 * Returns array of {original, lineIndex} objects.
 */
function extractUkrainianSegments(content) {
    const lines = content.split('\n');
    const segments = new Map(); // segment -> first lineIndex

    for (let i = 0; i < lines.length; i++) {
        const line = lines[i];
        if (!hasUkrainian(line)) continue;

        // Extract tokens: non-whitespace runs that contain Ukrainian
        const tokens = line.match(/\S+/g) || [];
        for (const token of tokens) {
            if (hasUkrainian(token) && !cache[token]) {
                if (!segments.has(token)) segments.set(token, i);
            }
        }

        // Also extract multi-word Ukrainian phrases (between quotes or in comments)
        const phrasePatterns = [
            /['"`]([^'"`]*[\u0400-\u04FF][^'"`]*?)['"`]/g,  // quoted strings
            /\/\/\s*(.*[\u0400-\u04FF].*)/g,                  // single-line comments
            /<!--\s*(.*[\u0400-\u04FF].*?)\s*-->/g,           // HTML comments
        ];
        for (const pat of phrasePatterns) {
            let m;
            while ((m = pat.exec(line)) !== null) {
                const phrase = m[1].trim();
                if (hasUkrainian(phrase) && phrase.length > 1 && !cache[phrase]) {
                    segments.set(phrase, i);
                }
            }
        }
    }

    return segments;
}

async function translateBatch(texts) {
    if (!texts.length) return;

    const numbered = texts.map((t, i) => `${i + 1}. ${t}`).join('\n');
    const prompt = `Translate the following Ukrainian words/phrases to European Portuguese (pt-PT).
Context: web application for elevator maintenance management.

Rules:
- Return ONLY a JSON object mapping original → translation
- Keep HTML tags, CSS class names, variable names, URLs unchanged
- Translate only human-readable Ukrainian text
- Use formal/professional Portuguese

Texts:
${numbered}

Return: {"original1": "translation1", "original2": "translation2", ...}`;

    try {
        const result = await model.generateContent(prompt);
        const text = result.response.text().trim();
        const jsonMatch = text.match(/\{[\s\S]*\}/);
        if (jsonMatch) {
            const translations = JSON.parse(jsonMatch[0]);
            for (const [orig, trans] of Object.entries(translations)) {
                if (trans && typeof trans === 'string') {
                    cache[orig] = trans;
                }
            }
        }
    } catch (e) {
        console.error(`  ⚠️ Gemini error: ${e.message}`);
    }
}

async function processFile(filepath) {
    let content;
    try {
        content = fs.readFileSync(filepath, 'utf-8');
    } catch (e) {
        console.log(`  ❌ Cannot read: ${e.message}`);
        return false;
    }

    if (!hasUkrainian(content)) return false;

    // Get segments to translate
    const segments = extractUkrainianSegments(content);
    const toTranslate = [...segments.keys()].filter(s => !cache[s]);

    if (toTranslate.length > 0) {
        console.log(`  🔄 Translating ${toTranslate.length} new segments...`);
        const BATCH = 40;
        for (let i = 0; i < toTranslate.length; i += BATCH) {
            await translateBatch(toTranslate.slice(i, i + BATCH));
            if (toTranslate.length > BATCH) await new Promise(r => setTimeout(r, 1500));
        }
        saveCache();
    }

    // Apply all cached translations to content
    let newContent = content;
    let changes = 0;

    // Sort by length descending to avoid partial matches
    const allToReplace = [...new Set([...segments.keys(), ...Object.keys(cache).filter(k => hasUkrainian(k) && content.includes(k))])];
    allToReplace.sort((a, b) => b.length - a.length);

    for (const original of allToReplace) {
        const translation = cache[original];
        if (!translation || translation === original || hasUkrainian(translation)) continue;
        if (newContent.includes(original)) {
            newContent = newContent.split(original).join(translation);
            changes++;
        }
    }

    if (changes > 0 && newContent !== content) {
        fs.writeFileSync(filepath, newContent, 'utf-8');
        return true;
    }

    return false;
}

async function main() {
    console.log('🇵🇹 Auto-translate Ukrainian → Portuguese (pt-PT)');
    console.log('='.repeat(60));

    const files = collectFiles(TARGET_DIRS);
    console.log(`📁 Found ${files.length} files\n`);

    let totalChanged = 0;

    for (const filepath of files.sort()) {
        const rel = filepath.replace('/home/andriy/deapseak/', '');

        // Quick pre-check
        try {
            const content = fs.readFileSync(filepath, 'utf-8');
            if (!hasUkrainian(content)) continue;
        } catch { continue; }

        console.log(`📄 ${rel}`);
        const changed = await processFile(filepath);

        if (changed) {
            console.log(`  ✅ File updated`);
            totalChanged++;
        } else {
            console.log(`  ℹ️  No changes`);
        }
    }

    // Final stats
    const remaining = { client: 0, tech: 0, dispatcher: 0 };
    for (const dir of TARGET_DIRS) {
        const key = dir.split('/')[1];
        try {
            const out = execSync(`grep -rl --include="*.html" --include="*.js" -P "[\\x{0400}-\\x{04FF}]" ${dir} 2>/dev/null | wc -l`, { encoding: 'utf-8' }).trim();
            remaining[key] = parseInt(out) || 0;
        } catch {}
    }

    console.log('\n' + '='.repeat(60));
    console.log(`✅ Done! Modified ${totalChanged} files`);
    console.log(`💾 Cache: ${Object.keys(cache).length} translations`);
    console.log(`\n📊 Files still with Ukrainian text:`);
    for (const [k, v] of Object.entries(remaining)) {
        console.log(`  ${k}: ${v} files`);
    }

    if (totalChanged > 0) {
        console.log('\n📝 To commit:');
        console.log('  git add pages/client pages/tech pages/dispatcher');
        console.log('  git commit -m "fix: translate Ukrainian → Portuguese in client/tech/dispatcher panels"');
    }
}

main().catch(console.error);
